File: Leetcode/1312 Minimum Insertion Steps to Make a String Palindrome.ts
function minInsertions(s: string): number {
    const cache = new Array<number[]>(s.length);
    for (let i = 0; i < cache.length; i++) {
        cache[i] = new Array<number>(s.length).fill(-1);
    }
    function minInserts(i: number, j: number): number {
        if (i > j) {
            return 0;
        }
        if (cache[i][j] !== -1) {
            return cache[i][j];
        }
        if (s[i] === s[j]) {
            cache[i][j] = minInserts(i + 1, j - 1);
        } else {
            cache[i][j] = 1 + Math.min(minInserts(i + 1, j), minInserts(i, j - 1));
        }
        return cache[i][j];
    }
    return minInserts(0, s.length - 1);
}

File: Leetcode/1312 Minimum Insertion Steps to Make a String Palindrome.py
class Solution:
    def minInsertions(self, s):
        n = len(s)
        cache = [[None] * n for _ in range(n)]
        def min_insertions(i, j):
            if i > j:
                return 0
            if cache[i][j] is not None:
                return cache[i][j]
            if s[i] == s[j]:
                cache[i][j] = min_insertions(i + 1, j - 1)
            else:
                cache[i][j] = 1 + min(min_insertions(i + 1, j), min_insertions(i, j - 1))
            return cache[i][j]
        return min_insertions(0, n - 1)
